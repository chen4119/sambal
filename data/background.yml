"@type": ImageObject
contentUrl: /data/images/image2.jpg
name: My background
author:
  "@id": /data/john smith.yml