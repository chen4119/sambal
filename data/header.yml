"@type": WPHeader
headline: Sambal
hasPart:
  - "@type": SiteNavigationElement
    name: About
    url: /about
  - "@type": SiteNavigationElement
    name: Docs
    url: /get-started
  - "@id": /collections/tags?tag=java%20script&output=sitenav
potentialAction:
  - "@type": "DownloadAction"
    target: https://github.com/chen4119/sambal