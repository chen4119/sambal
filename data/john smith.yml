"@type": Person
name: John Smith
email: smitty123@hotmail.com
familyName: Smith
givenName: John