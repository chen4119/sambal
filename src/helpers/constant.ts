import { URLSearchParams } from "url";

export type Theme = {
    name: string,
    options: object
};

export type PartitionKey = {[key: string]: string | number | Date};

export type Collection = {
    uri: string,
    include: string | string[],
    groupBy?: (mainEntity: any) => PartitionKey | PartitionKey[],
    sort?: (a: any, b: any) => number
};


export type WebPage = {
    "@type": string,
    url: string,
    mainEntity: any,
    [key: string]: any
};

export type OnBundleChanged = (isError: boolean, entry: any) => void;

export interface IHtmlSerializer {
    toHtml(renderResult: unknown): string;
};

export type URI = {
    protocol?: string,
    host?: string,
    path: string,
    query?: URLSearchParams
};

export interface IResolver {
    resolveUri(uri: URI): Promise<any>;
    clearCache(): void;
};

/*
export enum EntityType {
    Page,
    Data
};*/

export const CACHE_FOLDER = ".sambal";
export const OUTPUT_FOLDER = "public";
export const PAGES_FOLDER = "pages";
// export const DATA_FOLDER = "data";
export const SAMBAL_ENTRY_FILE = "sambal.entry.js";
export const SAMBAL_SITE_FILE = "sambal.site.js";

// export const MOUNT_FILE = "_mount.yml";
export const PAGE_FILE = "_page.yml";
export const LOCALHOST = "localhost";
export const FILE_PROTOCOL = "file:";

// for DevServer
export const THEME_PUBLIC_PATH = "/_theme";
export const DEV_PUBLIC_PATH = "/_sambal";

