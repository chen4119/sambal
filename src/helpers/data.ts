import axios, { AxiosResponse } from "axios";
import yaml from "js-yaml";
// import marked from "marked";
import glob from "glob";
import {
    getAbsFilePath,
    isValidRelativePath,
    readFileAsBuffer,
    readTextFile,
    frontMatter,
    safeParseJson,
    getFileExt
} from "./util";
import { isAbsUri } from "sambal-jsonld";

enum SUPPORTED_CONTENT_TYPE {
    yaml,
    markdown,
    json,
    image
}

export function searchFiles(query: string | string[]): string[] {
    if (Array.isArray(query)) {
        const matchSet = query.reduce((accumulator, value) => {
            searchFiles(value).forEach(m => accumulator.add(m));
            return accumulator;
        }, new Set<string>());
        return Array.from(matchSet);
    }
    if (isAbsUri(query)) {
        return [query];
    }
    const matches = glob.sync(query);
    return matches.filter(m => isValidRelativePath(m));
}

export function inferUrl(uri: string) {
    if (!uri || typeof(uri) !== "string") {
        return "";
    }
    if (uri.startsWith("/pages")) {
        let url = uri.substring(6);
        if (isSupportedFile(url)) {
            url = url.substring(0, url.lastIndexOf("."));
        }
        if (url === "/index") {
            return "/";
        }
        if (url.endsWith("/index")) {
            url = url.substring(0, url.length - 6);
        }
        return url;
    }
    return uri;
}

export async function loadLocalFile(src: string) {
    if (!isSupportedFile(src)) {
        throw new Error(`Unsupported file type: ${src}`);
    }

    const contentType = getLocalFileContentType(src);
    let content;
    if (contentType === SUPPORTED_CONTENT_TYPE.image) {
        content = await readFileAsBuffer(getAbsFilePath(src));
    } else {
        content = await readTextFile(getAbsFilePath(src));
    }
    return parseContent(content, contentType);
}

export async function loadRemoteFile(src: string) {
    const response = await axios.get(src);
    return parseContent(response.data, getAxiosResponseContentType(response));
}

const IMAGE_EXT_REGEX = /.+(.jpg|.jpeg|.gif|.png|.webp)$/i;
const SUPPORTED_FILE_EXT_REGEX = /.+(.yaml|.yml|.json|.md)$/i;

export function isImageFile(src: string) {
    if (!src) {
        return false;
    }
    return src.match(IMAGE_EXT_REGEX);
}

export function isSupportedFile(src: string) {
    if (!src) {
        return false;
    }
    return isAbsUri(src) || 
    isImageFile(src) ||
    src.match(SUPPORTED_FILE_EXT_REGEX);
}

function getAxiosResponseContentType(response: AxiosResponse<any>): SUPPORTED_CONTENT_TYPE {
    const contentType = response.headers["content-type"];
    if (contentType) {
        switch (contentType) {
            case "text/yaml":
            case "text/yml":
            case "application/yaml":
            case "application/yml":
                return SUPPORTED_CONTENT_TYPE.yaml;
            case "application/json":
            case "application/ld+json":
                return SUPPORTED_CONTENT_TYPE.json;
            case "text/markdown":
                return SUPPORTED_CONTENT_TYPE.markdown;
            case "image/jpeg":
            case "image/webp":
            case "image/gif":
            case "image/png":
                return SUPPORTED_CONTENT_TYPE.image;
        }
    }
    throw `Unsupported content type ${contentType}`;
}

function getLocalFileContentType(src: string): SUPPORTED_CONTENT_TYPE {
    const ext = getFileExt(src);
    switch (ext) {
        case "yaml":
        case "yml":
            return SUPPORTED_CONTENT_TYPE.yaml;
        case "json":
            return SUPPORTED_CONTENT_TYPE.json;
        case "md":
            return SUPPORTED_CONTENT_TYPE.markdown;
        case "jpg":
        case "jpeg":
        case "webp":
        case "gif":
        case "png":
            return SUPPORTED_CONTENT_TYPE.image;
        default:
            throw `Unsupported file type with extension ${ext}`;
    }
}

// TODO: ensure markdown is only for creative work!
function parseContent(content: any, contentType: SUPPORTED_CONTENT_TYPE) {
    try {
        let md;
        switch (contentType) {
            case SUPPORTED_CONTENT_TYPE.yaml:
                return yaml.load(content);
            case SUPPORTED_CONTENT_TYPE.json:
                return typeof(content) === "string" ? safeParseJson(content) : content;
            case SUPPORTED_CONTENT_TYPE.markdown:
                md = frontMatter(content);
                return {
                    ...md.frontMatter ? yaml.load(md.frontMatter) as object : {},
                    text: md.content,
                    encodingFormat: "text/markdown"
                }
            case SUPPORTED_CONTENT_TYPE.image:
                return content;
                // return Buffer.from(content, 'binary');
            default:
                throw `Unsupported content type ${contentType}.  Expecting yaml, json or markdown`;
        }
    } catch (e) {
        throw `Unable to parse content: ${e}`;
    }
}