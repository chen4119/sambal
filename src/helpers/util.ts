import fs from "fs";
import path from "path";
import shelljs from "shelljs";

export function isObjectLiteral(obj) {
    return obj !== null && typeof(obj) === "object" && Object.getPrototypeOf(obj) === Object.prototype;
}

export function isJsDate(value) {
    return typeof(value) === "object" && Object.getPrototypeOf(value) === Date.prototype;
}

export function isNullOrUndefined(val: any) {
    return typeof(val) === "undefined" || val === null;
}

export function getAbsFilePath(src: string) {
    return path.normalize(`${process.cwd()}/${src}`);
}

export function readTextFile(src: string): Promise<string> {
    return new Promise((resolve, reject) => {
        fs.readFile(src, "utf-8", (err, data) => {
            if (err) {
                reject(err);
            } else {
                resolve(data);
            }
        })
    });
}

export function readFileAsBuffer(src: string): Promise<any> {
    return new Promise((resolve, reject) => {
        fs.readFile(src, (err, data) => {
            if (err) {
                reject(err);
            } else {
                resolve(data);
            }
        })
    });
}

export function writeText(output: string, content: string): Promise<void> {
    return new Promise((resolve, reject) => {
        shelljs.mkdir("-p", path.dirname(output));
        fs.writeFile(output, content, "utf-8", (err) => {
            if (err) {
                reject(err);
            } else {
                resolve();
            }
        });
    });
}

export function writeBuffer(output: string, content: Buffer): Promise<void> {
    return new Promise((resolve, reject) => {
        shelljs.mkdir("-p", path.dirname(output));
        fs.writeFile(output, content, (err) => {
            if (err) {
                reject(err);
            } else {
                resolve();
            }
        });
    });
}

export function getFileExt(filePath: string) {
    const index = filePath.lastIndexOf(".");
    if (index >= 0 && index < filePath.length - 1) {
        return filePath.substring(index + 1).toLowerCase();
    }
    return "";
}

export function getMimeType(ext: string) {
    switch(ext) {
        case "js":
        case "mjs":
            return "text/javascript";
        case "ico":
            return "image/vnd.microsoft.icon";
        case "css":
            return `text/${ext}`;
        case "woff":
        case "woff2":
            return `font/${ext}`;
        case "jpeg":
        case "webp":
        case "gif":
        case "png":
            return `image/${ext}`;
        case "svg":
            return "image/svg+xml";
        default:
            throw new Error(`Unrecognized mime type ${ext}`);
    }
}

export function safeParseJson(jsonStr: string) {
    try {
        return JSON.parse(jsonStr);
    } catch (e) {
        return {};
    }
}

export function formatSize(size) {
    const base = Math.floor( Math.log(size) / Math.log(1024) );
    return `${Math.floor(size / Math.pow(1024, base))} ${['B', 'kB', 'MB', 'GB', 'TB'][base]}`;
}

// const FRONT_MATTER_REGEX = /---(.+)---(.+)/s
export function frontMatter(md: string) {
    const splitted = md.split("---");
    if (splitted.length >= 1) {
        return {
            frontMatter: splitted[1],
            content: splitted.slice(2).join("---")
        };
    }
    return {
        content: md
    }
}

/*
export async function traverseSiteGraph(root: PageNode, callback: (page: PageNode) => void | Promise<void>) {
    await callback(root);
    for (const child of root.hasPart) {
        await traverseSiteGraph(child, callback);
    }
}*/