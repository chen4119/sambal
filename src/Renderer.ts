import shelljs from "shelljs";
import ReactSerializer from "./ui/ReactSerializer";
import WebpackListenerPlugin from "./WebpackListenerPlugin";
import { replaceScriptSrc } from "./helpers/html";
import {
    bundleSambalFile,
    watchSambalFile,
    bundleBrowserPackage,
    getDevServerBrowserCompiler
} from "./helpers/bundler";
import {
    getAbsFilePath,
    isObjectLiteral,
    getMimeType,
    readFileAsBuffer,
    getFileExt
} from "./helpers/util";
import {
    CACHE_FOLDER,
    OUTPUT_FOLDER,
    SAMBAL_ENTRY_FILE,
    THEME_PUBLIC_PATH,
    DEV_PUBLIC_PATH,
    Theme,
    OnBundleChanged,
    IHtmlSerializer,
    WebPage
} from "./helpers/constant";
import { serializeJsonLd, renderSocialMediaMetaTags } from "./helpers/seo";
import { log } from "./helpers/log";

type UI = {
    renderPage: (props: {
        page: unknown,
        options?: unknown
    }) => Promise<unknown>,
    renderComponent: (props: {
        mainEntity: unknown,
        options?: unknown
    }) => Promise<unknown>,
    defaultOptions?: object,
    browserBundle?: {
        entry: unknown
    }
}

export default class Renderer {
    private serializer: IHtmlSerializer;
    private internalRenderer: UI;
    private internalBrowserBundleEntry: object;
    private themeRenderer: UI;
    private themeBrowserBundleEntry: object;
    private themeFolder: string;
    private themeOptions: object;

    constructor(
        private baseUrl: string,
        private entryFile: string,
        private theme: string | Theme) {
        this.serializer = new ReactSerializer();
        if (theme) {
            if (typeof(theme) === "string") {
                this.themeFolder = theme;
            } else if (isObjectLiteral(this.theme)) {
                this.themeFolder = theme.name;
                this.themeOptions = theme.options && isObjectLiteral(theme.options) ? theme.options : {};
            }
        }
    }

    async build(publicPath: string) {
        if (this.entryFile) {
            log.info("Bundling sambal.entry.js...");
            await bundleSambalFile(this.entryFile, getAbsFilePath(`${CACHE_FOLDER}/output`));
            this.internalRenderer = require(getAbsFilePath(`${CACHE_FOLDER}/output/${SAMBAL_ENTRY_FILE}`));
            if (this.internalRenderer.browserBundle) {
                log.info("Bundling browser bundle...");
                this.internalBrowserBundleEntry = await bundleBrowserPackage(
                    this.internalRenderer.browserBundle,
                    getAbsFilePath(`${OUTPUT_FOLDER}/${publicPath}`)
                );
            }
        }
        await this.initTheme();
        if (this.themeBrowserBundleEntry) {
            const outputDir = `${OUTPUT_FOLDER}/${publicPath}/${this.themeFolder}`;
            shelljs.mkdir("-p", outputDir);
            shelljs.cp("-R",
                getAbsFilePath(`${this.themeFolder}/dist/client/*`),
                getAbsFilePath(outputDir));
        }
        if (!this.internalRenderer && !this.themeRenderer) {
            throw new Error("No html renderer available.  Implement sambal.entry.js or specify a theme in sambal.site.js");
        }
    }

    async initTheme() {
        if (this.themeFolder) {
            try {
                log.info(`Loading theme ${this.themeFolder}`);
                const module = require(getAbsFilePath(`${this.themeFolder}/dist/index.js`));
                this.themeRenderer = module.entry;
                this.themeBrowserBundleEntry = module.browserBundle;
                return true;
            } catch (e) {
                log.error("Error loading theme", e);
            }
        }
        return false;
    }

    watchForEntryChange(onChange: OnBundleChanged) {
        if (this.entryFile) {
            return watchSambalFile(this.entryFile, (isError, entry) => {
                if (!isError) {
                    const module = require(getAbsFilePath(`${CACHE_FOLDER}/watch/${entry.main}`));
                    this.internalRenderer = module;
                    onChange(isError, entry);
                }
            });
        } else {
            onChange(false, {});
            return null;
        }
    }

    watchForBrowserBundleChange(onChange: OnBundleChanged) {
        if (this.internalRenderer && this.internalRenderer.browserBundle) {
            const listener = new WebpackListenerPlugin((isError, entry) => {
                this.internalBrowserBundleEntry = entry;
                onChange(isError, entry);
            });
            return getDevServerBrowserCompiler(this.internalRenderer.browserBundle, listener);
        }
        return null;
    }

    async renderPage(page: WebPage, publicPath?: string) {
        let renderResult;
        let clientBundle;
        let bundlePrefix;
        if (this.internalRenderer) {
            const defaultOptions = this.getDefaultOptions(this.internalRenderer);
            renderResult = await this.internalRenderer.renderPage({ 
                page: page,
                options: defaultOptions
            });
            clientBundle = this.internalBrowserBundleEntry;
            bundlePrefix = publicPath ? publicPath : DEV_PUBLIC_PATH;
        }
        // if internalRenderer didn't render, try theme renderer, if available
        if (!renderResult && this.themeRenderer) {
            const defaultOptions = this.getDefaultOptions(this.themeRenderer);
            renderResult = await this.themeRenderer.renderPage({ 
                page: page,
                options: {
                    ...defaultOptions,
                    ...this.themeOptions
                }
            });
            clientBundle = this.themeBrowserBundleEntry;
            bundlePrefix = publicPath ? `${publicPath}/${this.themeFolder}` : THEME_PUBLIC_PATH;
        }
        if (renderResult) {
            const html = typeof(renderResult) === "string" ? 
                renderResult :
                this.serializer.toHtml(renderResult);
            return await this.postProcessHtml(page, html, clientBundle, bundlePrefix);
        }
        return null;
    }

    private async postProcessHtml(page: WebPage, html: string, bundle: object, prefix: string) {
        if (!bundle) {
            return html;
        }
        let hasJsonLd = false;
        let hasSocialMediaMeta = false;
        let updatedHtml = await replaceScriptSrc(html, (name, attribs) => {
            if (name === "script" && attribs.type === "application/ld+json") {
                hasJsonLd = true;
            } else if (name === "meta" && this.isSocialMediaMeta(attribs.name)) {
                hasSocialMediaMeta = true;
            }

            if (name === "script" && attribs.src) {
                let realSrc = attribs.src;
                for (const entryName of Object.keys(bundle)) {
                    if (attribs.src === entryName) {
                        realSrc = `${prefix}/${bundle[entryName]}`;
                        break;
                    }
                }
                return {
                    ...attribs,
                    src: realSrc
                };
            }
            return attribs;
        });
        if (!hasSocialMediaMeta && page.mainEntity) {
            updatedHtml = await this.addSocialMediaMeta(updatedHtml, page.mainEntity);
        }
        if (!hasJsonLd && page.mainEntity) {
            updatedHtml = this.addJsonLdScript(updatedHtml, page.mainEntity);
        }
        return updatedHtml;
    }

    private isSocialMediaMeta(name: string) {
        return name && (name.startsWith("og:") || name.startsWith("twitter:"));
    }

    private async addSocialMediaMeta(html: string, mainEntity: any) {
        const metaTags = await renderSocialMediaMetaTags(this.baseUrl, mainEntity);

        const index = html.indexOf("<head>");
        if (index >= 0) {
            return html.substring(0, index + 6) + metaTags + html.substring(index + 6);
        }
        return html;
    }

    private addJsonLdScript(html: string, mainEntity: any) {
        const serializedJsonLd = serializeJsonLd(mainEntity);

        const jsonLdScript = `
        <script type="application/ld+json">
            ${serializedJsonLd}
        </script>`;

        const index = html.indexOf("</head>");
        if (index >= 0) {
            return html.substring(0, index) + jsonLdScript + html.substring(index);
        }
        return html;
    }

    async getThemeFile(filePath: string) {
        return {
            mime: getMimeType(getFileExt(filePath)),
            data: await readFileAsBuffer(getAbsFilePath(`${this.themeFolder}/dist/client/${filePath}`))
        }
    }

    private getDefaultOptions(renderer: UI) {
        if (renderer.defaultOptions && isObjectLiteral(renderer.defaultOptions)) {
            return renderer.defaultOptions;
        }
        return {};
    }
}