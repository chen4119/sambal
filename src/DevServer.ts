import express from "express";
// import { Watching } from "webpack";
import webpackDevMiddleware from "webpack-dev-middleware";
import Renderer from "./Renderer";
import { WebPage, THEME_PUBLIC_PATH, DEV_PUBLIC_PATH } from "./helpers/constant";
import { log } from "./helpers/log";
import Router from "./Router";

export default class DevServer {
    private expressApp;
    private server;
    // private watchEntryFile: Watching;
    
    constructor(private router: Router, private renderer: Renderer, private port: Number) {
        
    }
    
    start() {
        this.renderer.watchForEntryChange((isError) => {
            log.info("sambal.entry.js compiled");
            if (!isError && !this.expressApp) {
                this.startDevServer();
            }
        });
    }

    stop() {
        this.server.close();
    }
    
    private startDevServer() {
        this.expressApp = express();
        this.addBrowserBundleMiddleware();
        this.expressApp.get(`${THEME_PUBLIC_PATH}/*`, this.getThemeFile.bind(this));
        this.expressApp.get("*", this.route.bind(this));
        this.server = this.expressApp.listen(this.port, () => {
            log.info(`Dev server started on port ${this.port}`);
        });
    }

    private addBrowserBundleMiddleware() {
        const compiler = this.renderer.watchForBrowserBundleChange(this.onBrowserBundleChanged.bind(this));
        if (compiler) {
            const middleware = webpackDevMiddleware(
                compiler, {
                    publicPath: DEV_PUBLIC_PATH
                });
            this.expressApp.use(middleware);
        }
    }

    private onBrowserBundleChanged(isError, entry) {
        log.info("Browser bundle compiled");
    }

    private async getThemeFile(req, res) {
        try {
            const file = await this.renderer.getThemeFile(req.path.substring(THEME_PUBLIC_PATH.length));
            res.set("Content-Type", file.mime);
            res.send(file.data);
        } catch (e) {
            log.error(e);
            res.status(404).end();
        }
    }

    private async route(req, res) {
        log.debug(`Get path uri: ${req.path}`);
        const page = await this.router.getPage(req.path);
        if (page) {
            let html = await this.renderer.renderPage(page);
            res.send(html);
        } else {
            res.status(404).end();
        }
    }
}