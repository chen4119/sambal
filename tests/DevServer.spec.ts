import shelljs from "shelljs";
import axios, { AxiosResponse } from "axios";
import Renderer from "../src/Renderer";
import { init, wait } from "./setup";
import DevServer from "../src/DevServer";
import { PAGES_FOLDER } from "../src/helpers/constant";
import { writeText, getAbsFilePath } from "../src/helpers/util";
import { loadLocalFile } from "../src/helpers/data";

const originalTestBlog = `
---
"@type": BlogPosting
headline: Test blog
---

Hello world3
`;

const modifiedTestBlog = `
---
"@type": BlogPosting
headline: Test blog
---

Hello world 3 changed!
`;

describe("DevServer", () => {
    let renderer: Renderer;
    let server: DevServer;
    const baseUrl = "https://example.com";
    const testFile = getAbsFilePath(`${PAGES_FOLDER}/testblog.md`);

    beforeAll(async () => {
        await writeText(testFile, originalTestBlog);
        const classes = await init();
        renderer = new Renderer(baseUrl, null, "mock-theme");
        await renderer.initTheme();
        server = new DevServer(classes.uriResolver, classes.media, classes.router, renderer, 3000);
        await server.start();
    });

    afterAll(async () => {
        shelljs.rm("-f", testFile);
        await server.stop();
    })

    it('render html', async () => {
        const response = await axios.get("http://localhost:3000");
        expect(response.data).toMatchSnapshot();
    });

    it('get client bundle', async () => {
        const response = await axios.get("http://localhost:3000/_theme/client.123.js");
        expect(response.data).toMatchSnapshot();
    });

    it('refresh data on file change', async () => {
        let response = await axios.get("http://localhost:3000/testblog");
        await writeText(testFile, modifiedTestBlog);
        await wait();
        response = await axios.get("http://localhost:3000/testblog");
        expect(response.data).toMatchSnapshot();
    });

    it('get original image2', async () => {
        const response = await axios.get("http://localhost:3000/data/images/image2.jpg");
        expect(response.status).toBe(200);
    });

    it('get image2 as webp and thumbnail', async () => {
        // load blog2 will trigger transformation of images
        let response = await axios.get("http://localhost:3000/blogs/blog2");

        response = await axios.get("http://localhost:3000/data/images/image2.webp");
        expect(response.status).toBe(200);
        response = await axios.get("http://localhost:3000/data/images/image2-50.webp");
        expect(response.status).toBe(200);
    });

});