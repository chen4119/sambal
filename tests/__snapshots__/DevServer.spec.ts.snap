// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevServer get client bundle 1`] = `"console.log(\\"Hello from client!\\");"`;

exports[`DevServer refresh data on file change 1`] = `
"
                <html>
                    <head>
            
        <link rel=\\"canonical\\" href=\\"https://example.com/testblog\\" />
        <title>Test blog</title>
    
            
        <meta name=\\"twitter:card\\" content=\\"summary\\" />
        <meta name=\\"twitter:title\\" content=\\"Test blog\\" />
        
    
            
        <meta name=\\"og:url\\" content=\\"https://example.com/testblog\\" />
        <meta name=\\"og:title\\" content=\\"Test blog\\" />
        
    
         
                        <script src=\\"/_theme/client.123.js\\"></script>
                    
        <script type=\\"application/ld+json\\">
            {
    \\"@type\\": \\"BlogPosting\\",
    \\"headline\\": \\"Test blog\\",
    \\"text\\": \\"\\\\n\\\\nHello world 3 changed!\\\\n\\",
    \\"encodingFormat\\": \\"text/markdown\\"
}
        </script></head>
                    <body>
                        <h1>Hello from mock theme</h1>
                    
        <script>
            function openSocket() {
                const sambalws = new WebSocket(\\"ws://localhost:3001/\\");
                sambalws.onopen = function() {
                    console.log('Sambal dev server connected');
                };
                
                sambalws.onmessage = function(e) {
                    if (e.data === \\"refresh\\") {
                        console.log('Reloading...');
                        window.location.reload();
                    }
                }
    
                sambalws.onclose = function(e) {
                    console.log(\\"Sambal dev server disconnected\\");
                    setTimeout(() => {
                        openSocket();
                    }, 2000);
                };
            }
            openSocket();
        </script></body>
                </html>
            "
`;

exports[`DevServer render html 1`] = `
"
                <html>
                    <head>
            
        <link rel=\\"canonical\\" href=\\"https://example.com/\\" />
        <title>A linked data static site generator</title>
    
            
        <meta name=\\"twitter:card\\" content=\\"summary\\" />
        <meta name=\\"twitter:title\\" content=\\"A linked data static site generator\\" />
        <meta name=\\"twitter:description\\" content=\\"A linked data static site generator\\" />
    
            
        <meta name=\\"og:url\\" content=\\"https://example.com/\\" />
        <meta name=\\"og:title\\" content=\\"A linked data static site generator\\" />
        <meta name=\\"og:description\\" content=\\"A linked data static site generator\\" />
    
         
                        <script src=\\"/_theme/client.123.js\\"></script>
                    
        <script type=\\"application/ld+json\\">
            {
    \\"@type\\": \\"SoftwareSourceCode\\",
    \\"name\\": \\"Sambal\\",
    \\"headline\\": \\"A linked data static site generator\\",
    \\"description\\": \\"A linked data static site generator\\",
    \\"text\\": \\"&lt;need to &lt;&gt; escape&gt;\\"
}
        </script></head>
                    <body>
                        <h1>Hello from mock theme</h1>
                    
        <script>
            function openSocket() {
                const sambalws = new WebSocket(\\"ws://localhost:3001/\\");
                sambalws.onopen = function() {
                    console.log('Sambal dev server connected');
                };
                
                sambalws.onmessage = function(e) {
                    if (e.data === \\"refresh\\") {
                        console.log('Reloading...');
                        window.location.reload();
                    }
                }
    
                sambalws.onclose = function(e) {
                    console.log(\\"Sambal dev server disconnected\\");
                    setTimeout(() => {
                        openSocket();
                    }, 2000);
                };
            }
            openSocket();
        </script></body>
                </html>
            "
`;
